src/providers/LedgerProvider.js

```diff
@@ -42,18 +43,31 @@ export default class LedgerProvider extends Provider {
     let path = false
     let index = 0
 
-    while (!path) {
+    // A maximum number of addresses to lookup after which it is deemed
+    // that the wallet does not contain this address
+    const maxAddresses = 50
+
+    while (!path && index < maxAddresses) {
       const addr = await this.getAddresses(index, 1)
       if (String(addr[0]) === address) path = this.getDerivationPathFromIndex(index)
       index++
     }
 
+    if (!path) {
+      throw new Error('Ledger: Wallet does not contain address')
+    }
+
     return path
   }
 
   async getAddressFromIndex (addressIndex) {
     const path = this.getDerivationPathFromIndex(addressIndex)
-    return this.getAddressFromDerivationPath(path)
+    if (path in this._addressCache) {
```

Comment:
@harshjv  I eventually didn't cache it by first address. This is because it would then have to be implemented in `getAddresses`. Unfortunately though, we are using `getAddresses(0,1)` everywhere and always looking up the first address in this function would remove any performance benefit. 
We can hopefully mandate somehow that another device needs a new instance of the provider.